libs/langchain-community/src/vectorstores/upstash.ts

```diff
@@ -0,0 +1,241 @@
+import * as uuid from "uuid";
+import { EmbeddingsInterface } from "@langchain/core/embeddings";
+import { VectorStore } from "@langchain/core/vectorstores";
+import { Index as UpstashIndex } from "@upstash/vector";
+import { Document } from "@langchain/core/documents";
+import { chunkArray } from "@langchain/core/utils/chunk_array";
+import {
+  AsyncCaller,
+  AsyncCallerParams,
+} from "@langchain/core/utils/async_caller";
+
+/**
+ * This interface defines the arguments for the UpstashVectorStore class.
+ */
+export interface UpstashVectorLibArgs extends AsyncCallerParams {
+  index: UpstashIndex;
+}
+
+// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
+export type UpstashMetadata = Record<string, any>;
+
+// eslint-disable-next-line @typescript-eslint/ban-types, @typescript-eslint/no-explicit-any
+export type UpstashQueryMetadata = UpstashMetadata & { documentContentLC: any };
+
+/**
+ * Type that defines the parameters for the delete method.
+ * It can either contain the target id(s) or the deleteAll config to reset all the vectors.
+ */
+export type UpstashDeleteParams =
+  | {
+      ids: string | string[];
+      deleteAll?: never;
+    }
+  | { deleteAll: boolean; ids?: never };
+
+const CONCURRENT_UPSERT_LIMIT = 1000;
+/**
+ * The main class that extends the 'VectorStore' class. It provides
+ * methods for interacting with Upstash index, such as adding documents,
+ * deleting documents, performing similarity search and more.
+ */
+export class UpstashVectorStore extends VectorStore {
+  index: UpstashIndex;
+
+  caller: AsyncCaller;
+
+  _vectorstoreType(): string {
+    return "upstash";
+  }
+
+  constructor(embeddings: EmbeddingsInterface, args: UpstashVectorLibArgs) {
+    super(embeddings, args);
+
+    this.embeddings = embeddings;
+
+    const { index, ...asyncCallerArgs } = args;
+
+    this.index = index;
+    this.caller = new AsyncCaller(asyncCallerArgs);
+  }
+
+  /**
+   * This method adds documents to Upstash database. Documents are first converted to vectors
+   * using the provided embeddings instance, and then upserted to the database.
+   * @param documents Array of Document objects to be added to the database.
+   * @param options Optional object containing array of ids for the documents.
+   * @returns Promise that resolves with the ids of the provided documents when the upsert operation is done.
+   */
+  async addDocuments(documents: Document[], options?: { ids?: string[] }) {
```

Comment:
Should be able to use `DocumentInterface` here, also update the other methods to reflect.